### ISSUE TITLE
[@typescript-eslint/no-unnecessary-condition] False positive "Unnecessary optional chain on a non-nullish value"

### ISSUE THREAD

**Issue by niklasf** [typescript-eslint/typescript-eslint#2247]
**Repro**

`.eslintrc.json`

```JSON
{
  "parser": "@typescript-eslint/parser",
  "extends": ["plugin:@typescript-eslint/recommended"],
  "parserOptions": {
    "sourceType": "module",
    "project": "./tsconfig.json"
  },
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": ["warn"]
  },
  "root": true
}
```

`tsconfig.json`
```JSON
{
  "compilerOptions": {
    "strict": true
  }
}
```

`src/test.ts`
```TS
export interface Outer {
  inner?: {
    [key: string]: string | undefined;
  }
}

export function test(outer: Outer, key: string): number | undefined {
  return outer.inner?.[key]?.charCodeAt(0);
}
```

**Expected Result**

No error or warning.

**Actual Result**

```
[...]/src/test.ts
  8:28  warning  Unnecessary optional chain on a non-nullish value  @typescript-eslint/no-unnecessary-condition

✖ 1 problem (0 errors, 1 warning)
  0 errors and 1 warning potentially fixable with the `--fix` option.
```

**Additional Info**

Running `--fix` changes the function to

```TS
export function test(outer: Outer, key: string): number | undefined {
  return outer.inner?.[key].charCodeAt(0);
}
```

but this is now a Typescript error:

```
src/test.ts:8:10 - error TS2532: Object is possibly 'undefined'.

8   return outer.inner?.[key].charCodeAt(0);
           ~~~~~~~~~~~~~~~~~~


Found 1 error.
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.4.0` |
| `@typescript-eslint/parser`        | `3.4.0` |
| `TypeScript`                       | `3.9.5` |
| `ESLint`                           | `7.3.1` |
| `node`                             | `14.4.0` |
| `npm`                              | `6.14.5` |

**Comment by bradzacher** [comment]
Took me a minute to find this bug.

The issue is as follows.

This line here only asserts that the property is an `Identifier`.
https://github.com/typescript-eslint/typescript-eslint/blob/5fdd21a1726fb6928098c4152aec55a30df960d4/packages/eslint-plugin/src/rules/no-unnecessary-condition.ts#L451

Then here it attempts to fetch the property type from the base type.
https://github.com/typescript-eslint/typescript-eslint/blob/5fdd21a1726fb6928098c4152aec55a30df960d4/packages/eslint-plugin/src/rules/no-unnecessary-condition.ts#L453

The problem is that in this instance, `property` is an `Identifier` (it's the variable `key`).
But because it's being used in a computed context, its name shouldn't be looked up as a property name in the parent - instead we should grab its type and use that to figure out how to get the relevant property from the parent (in this case, it's got a `string` type, so it should instead grab the string index signature).

**Comment by niklasf** [comment]
Yay, thanks :)
